### PR TITLE
Avoid conflicts when installing packages from other sources

### PR DESCRIPTION
**Pull request by nanzhg**
To install dependent packages from other sources for testing use, it requires to allow installing renamed packages.
https://openqa.suse.de/tests/15782859#step/kubevirt_tests_server/78

- Related ticket: https://progress.opensuse.org/issues/168568
- Verification run: https://openqa.suse.de/tests/15811368

### CODE REVIEW

**Comment by alice-suse** [comment]
LGTM, let's wait for validation.

**Comment by waynechen55** [comment]
Please help provide evidence that your installed packages will not modify SL-Micro packages.

**Comment by waynechen55** [comment]
> Please help provide evidence that your installed packages will not modify SL-Micro packages.

It looks to me you replaces SL Micro package perl-Bootloader with update-bootloader from other repos.

**Comment by nanzhg** [comment]
> > Please help provide evidence that your installed packages will not modify SL-Micro packages.
> 
> It looks to me you replaces SL Micro package perl-Bootloader with update-bootloader from other repos.

Yes, it was automatically processed by zypper. Required package 'perl-CPAN-Changes' during testing depends on this renamed package 'update-bootloader'.

**Comment by waynechen55** [comment]
> > > Please help provide evidence that your installed packages will not modify SL-Micro packages.
> > 
> > 
> > It looks to me you replaces SL Micro package perl-Bootloader with update-bootloader from other repos.
> 
> Yes, it was automatically processed by zypper. Required package 'perl-CPAN-Changes' during testing depends on this renamed package 'update-bootloader'.

I do not think it is good, namely replacing SL Micro package with non-SL Micro package. Please help considering avoid this.

File: lib/registration.pm [hunk 1]
```diff
@@ -1075,7 +1075,10 @@ sub runtime_registration {
     if (is_transactional) {
         trup_call('register' . $cmd);
         trup_call('--continue run zypper --gpg-auto-import-keys refresh') if is_staging;
-        add_suseconnect_product('SL-Micro-Extras') if (is_sle_micro('>=6.0'));
+        if (is_sle_micro('>=6.0')) {
+            process_reboot(trigger => 1);
+            add_suseconnect_product('SL-Micro-Extras');
+        }
```
Comment (waynechen55) [inline lib/registration.pm:1081]: Why make this change ?

Comment (nanzhg) [reply lib/registration.pm:1081]: This code snippet is from another PR which mentioned in related ticket, it's just for testing use.

**Comment by nanzhg** [comment]
> LGTM, let's wait for validation.

Verified this code change will fix the problem - https://openqa.suse.de/tests/15801867

**Comment by volodymyrkatkalov** [comment]
Has duplications with https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20494

File: lib/registration.pm (hunk 1 above)
Comment (alice-suse) [reply lib/registration.pm:1081]: @nanzhg The  commit is already given in another PR, please exclude this change and only include yours before merge.

**Comment by alice-suse** [comment]
> > > > Please help provide evidence that your installed packages will not modify SL-Micro packages.
> > > 
> > > 
> > > It looks to me you replaces SL Micro package perl-Bootloader with update-bootloader from other repos.
> > 
> > 
> > Yes, it was automatically processed by zypper. Required package 'perl-CPAN-Changes' during testing depends on this renamed package 'update-bootloader'.
> 
> I do not think it is good, namely replacing SL Micro package with non-SL Micro package. Please help considering avoiding this. You can install a lower version by appending version to package name. Also check whether SL Micro already provides similar package (they may have different name in SL Micro repo). Or consider whether you really needs perl-CPAN-Changes or other alternative tools available.

@nanzhg How do you think with above? 

Basically I agree, using rpms from Packages which is supported by suse will be better than using rpms from Extras or other user added repo which is NOT SUPPORTED (only provided) by suse.

perl-CPAN-Changes does not seem a critical pkg, is it necessary in kubevirt test? If not , can removing it  solve the problem with low effort?

Otherwise, in worst case, the proposal change affects pkgs in "INSTALL_OTHER_PACKAGES" which are basically automation needed packages. So impact to test system will be lower, so acceptable for me,although not ideal.

File: tests/virt_autotest/prepare_transactional_server.pm [hunk 2]
```diff
@@ -86,7 +86,7 @@ sub install_additional_pkgs {
         zypper_call("--gpg-auto-import-keys refresh");
         save_screenshot;
 
-        my $cmd = "install --no-allow-downgrade --no-allow-name-change --no-allow-vendor-change";
+        my $cmd = "install --no-allow-downgrade --no-allow-vendor-change";
```
Comment (waynechen55) [inline tests/virt_autotest/prepare_transactional_server.pm:89]: See my comment [1](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20498#issuecomment-2443721359) and [2](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20498#issuecomment-2448853946).

**Review by waynechen55 — CHANGES_REQUESTED** [review]

**Comment by nanzhg** [comment]
> > I do not think it is good, namely replacing SL Micro package with non-SL Micro package. Please help considering avoiding this. You can install a lower version by appending version to package name. Also check whether SL Micro already provides similar package (they may have different name in SL Micro repo). Or consider whether you really needs perl-CPAN-Changes or other alternative tools available.
> 
> @nanzhg How do you think with above?
> 
> Basically I agree, using rpms from Packages which is supported by suse will be better than using rpms from Extras or other user added repo which is NOT SUPPORTED (only provided) by suse.
> 
> perl-CPAN-Changes does not seem a critical pkg, is it necessary in kubevirt test? If not , can removing it solve the problem with low effort?
> 
> Otherwise, in worst case, the proposal change affects pkgs in "INSTALL_OTHER_PACKAGES" which are basically automation needed packages. So impact to test system will be lower, so acceptable for me,although not ideal.

For the package 'perl-CPAN-Changes', I can't find it in sle-micro repos. And also it will cause other dependencies issue while installing 'ant-junit' package, both are for generating XML/HTML report use after all tests done. As I tried, only Tumbleweed repository can be used for resolve the issue.
Currently, I think this might be a temporarily solution. As we have another ticket to improve the report, maybe use a new framework which does not depend on that. If we have the new solution to solve it, then we can change it back.

```
transactional update ~# zypper lr -E
Repository priorities are without effect. All enabled repositories share the same priority.

# | Alias                                                 | Name                     | Enabled | GPG Check | Refresh
--+-------------------------------------------------------+--------------------------+---------+-----------+--------
1 | Basesystem-15-SP5                                     | Basesystem-15-SP5        | Yes     | (r ) Yes  | No
2 | Basesystem-15-SP6                                     | Basesystem-15-SP6        | Yes     | (r ) Yes  | No
3 | Development-Tools-15-SP5                              | Development-Tools-15-SP5 | Yes     | (r ) Yes  | No
4 | Development-Tools-15-SP6                              | Development-Tools-15-SP6 | Yes     | (r ) Yes  | No
5 | SUSE_Linux_Extras_6.0_x86_64:SL-Micro-Extras-6.0-Pool | SL-Micro-Extras-6.0-Pool | Yes     | (r ) Yes  | Yes
8 | SUSE_Linux_Micro_6.0_x86_64:SL-Micro-6.0-Pool         | SL-Micro-6.0-Pool        | Yes     | (r ) Yes  | No

transactional update ~# zypper in make perl-CPAN-Changes ant-junit git bind-utils
Refreshing service 'SUSE_Linux_Extras_6.0_x86_64'.
Refreshing service 'SUSE_Linux_Micro_6.0_x86_64'.
Loading repository data...
Reading installed packages...
Package 'perl-CPAN-Changes' not found.
Resolving package dependencies...

Problem: nothing provides 'xerces-j2' needed by the to be installed ant-1.10.14-150200.4.18.2.noarch
 Solution 1: do not install ant-junit-1.10.14-150200.4.28.1.noarch
 Solution 2: break ant-1.10.14-150200.4.18.2.noarch by ignoring some of its dependencies

Choose from above solutions by number or cancel [1/2/c/d/?] (c):
```

**Comment by nanzhg** [comment]
> Has duplications with #20494

Yes, it's just for testing. Will remove it in final version. Thanks

File: lib/registration.pm (hunk 1 above)
Comment (nanzhg) [reply lib/registration.pm:1081]: Sure, will remove it in final commit. For now, I'd like to decide how to resolve the package install conflicts.

**Comment by alice-suse** [comment]
> > > I do not think it is good, namely replacing SL Micro package with non-SL Micro package. Please help considering avoiding this. You can install a lower version by appending version to package name. Also check whether SL Micro already provides similar package (they may have different name in SL Micro repo). Or consider whether you really needs perl-CPAN-Changes or other alternative tools available.
> > 
> > 
> > @nanzhg How do you think with above?
> > Basically I agree, using rpms from Packages which is supported by suse will be better than using rpms from Extras or other user added repo which is NOT SUPPORTED (only provided) by suse.
> > perl-CPAN-Changes does not seem a critical pkg, is it necessary in kubevirt test? If not , can removing it solve the problem with low effort?
> > Otherwise, in worst case, the proposal change affects pkgs in "INSTALL_OTHER_PACKAGES" which are basically automation needed packages. So impact to test system will be lower, so acceptable for me,although not ideal.
> 
> For the package 'perl-CPAN-Changes', I can't find it in sle-micro repos. And also it will cause other dependencies issue while installing 'ant-junit' package, both are for generating XML/HTML report after all tests done. As I tried, only Tumbleweed repository can be used for resolve the issue. Currently, I think this might be a temporarily solution. As we have another ticket to improve the report, there will be a new solution to replace the existing one which does not depend on that. If we have the new solution to solve it, then we can change it back.
> 
> ```
> transactional update ~# zypper lr -E
> Repository priorities are without effect. All enabled repositories share the same priority.
> 
> # | Alias                                                 | Name                     | Enabled | GPG Check | Refresh
> --+-------------------------------------------------------+--------------------------+---------+-----------+--------
> 1 | Basesystem-15-SP5                                     | Basesystem-15-SP5        | Yes     | (r ) Yes  | No
> 2 | Basesystem-15-SP6                                     | Basesystem-15-SP6        | Yes     | (r ) Yes  | No
> 3 | Development-Tools-15-SP5                              | Development-Tools-15-SP5 | Yes     | (r ) Yes  | No
> 4 | Development-Tools-15-SP6                              | Development-Tools-15-SP6 | Yes     | (r ) Yes  | No
> 5 | SUSE_Linux_Extras_6.0_x86_64:SL-Micro-Extras-6.0-Pool | SL-Micro-Extras-6.0-Pool | Yes     | (r ) Yes  | Yes
> 8 | SUSE_Linux_Micro_6.0_x86_64:SL-Micro-6.0-Pool         | SL-Micro-6.0-Pool        | Yes     | (r ) Yes  | No
> 
> transactional update ~# zypper in make perl-CPAN-Changes ant-junit git bind-utils
> Refreshing service 'SUSE_Linux_Extras_6.0_x86_64'.
> Refreshing service 'SUSE_Linux_Micro_6.0_x86_64'.
> Loading repository data...
> Reading installed packages...
> Package 'perl-CPAN-Changes' not found.
> Resolving package dependencies...
> 
> Problem: nothing provides 'xerces-j2' needed by the to be installed ant-1.10.14-150200.4.18.2.noarch
>  Solution 1: do not install ant-junit-1.10.14-150200.4.28.1.noarch
>  Solution 2: break ant-1.10.14-150200.4.18.2.noarch by ignoring some of its dependencies
> 
> Choose from above solutions by number or cancel [1/2/c/d/?] (c):
> ```

I see. To me, I can accept current PR as temporary workaround. Let's change it back once ant is not needed which will be about 1 month away.

File: tests/virt_autotest/prepare_transactional_server.pm (hunk 2 above)
Comment (nanzhg) [reply tests/virt_autotest/prepare_transactional_server.pm:89]: Another solution is that handling the package installation in kubevirt test code.

Comment (waynechen55) [reply tests/virt_autotest/prepare_transactional_server.pm:89]: Yes. I prefer you handle in your own test. So other tests will not have this risky code.

File: tests/virt_autotest/kubevirt_tests_server.pm [hunk 3]
```diff
@@ -237,8 +236,21 @@ sub install_kubevirt_packages {
     zypper_call('in jq open-iscsi') if (script_run('rpmquery jq open-iscsi') && ($kubevirt_ver ge "0.50.0"));
 
     # Install required packages perl-CPAN-Changes and ant-junit
+    my $repo_name = '';
+    my $inst_cmd = 'in -f';
```
Comment (waynechen55) [inline tests/virt_autotest/kubevirt_tests_server.pm:240]: If these two variables ```repo_name``` and ```inst_cmd``` are only used inside if block, can they moved into there ?

Comment (nanzhg) [reply tests/virt_autotest/kubevirt_tests_server.pm]: Thanks for review! I made relative changes. For the variable `repo_name`, it requires to generate random number for each repository name as per the code logic.

**Review by waynechen55 — APPROVED** [review]

File: tests/virt_autotest/kubevirt_tests_server.pm [hunk 4]
```diff
@@ -237,8 +236,21 @@ sub install_kubevirt_packages {
...
+        if (get_var("INSTALL_OTHER_REPOS")) {
+            foreach (split(/,/, get_var("INSTALL_OTHER_REPOS"))) {
+                $repo_name = (split(/\//, $_))[-1] . "-" . bmwqemu::random_string(8);
+                zypper_call("--gpg-auto-import-keys ar --enable --refresh $_ $repo_name");
+                save_screenshot;
+            }
+            zypper_call("--gpg-auto-import-keys refresh");
+            save_screenshot;
+            $inst_pkgs = $inst_pkgs . " $_" foreach (split(/,/, get_required_var("INSTALL_OTHER_PACKAGES")));
+            zypper_call('in -f' . $inst_pkgs);
```
Comment (alice-suse) [inline tests/virt_autotest/kubevirt_tests_server.pm:251]: Would you please also remove the repo after installing the assistant pkgs? BTW, a comment describing that this is temporary solution and will be removed after ant is not replaced by other way in roadmap 18, will be helpful. Thanks.

Comment (alice-suse) [reply tests/virt_autotest/kubevirt_tests_server.pm:251]: SLM increments are failing by missing this. So merge it now. Please improvethis minor thing in new PR.

**Pull request merged**